File: android/src/com/panyam/mango/templates/filters/Filter.java
package com.panyam.mango.templates.filters;

import java.io.*;
import java.util.*;
import com.panyam.mango.templates.variables.*;
import com.panyam.mango.templates.core.*;

/**
 * Filters get applied on variables in how they manipulate the value stored in a variable.
 * 
 * We do not want to have multiple instances of the same filter.  So it would be better to 
 * have a single instance of the filter through out the application to save memory and their
 * performance wouldnt be bad either.  Since only a certain number of filter instances are 
 * created to conserve memory, this implies that Filter objects cannot have shared state or 
 * state specific to each invocation(of their render) method.  This is for the sake of thread
 * safety.
 *
 * So as an extension (at a later date), this restriction on shared or invocation specific 
 * state can be relaxed by having a FilterContext object associated with each filter if
 * this is a paramount requirement.
 * 
 * @author Sri Panyam
 */
public class Filter
{
    /**
     * Creates a new filter.
     */
    public Filter()
    {
    }
    
    /**
     * Creates a filter instance specific context.
     * @return A filter specific context object.
     */
    public Object makeInstanceContext(ArrayList<Variable> arguments)
    {
        return null;
    }
    
    /**
     * Creates an invocation specific context for the filter.
     * @return An invocation specific context object for this filter.
     */
    public Object makeInvocationContext(ArrayList<Variable> arguments)
    {
        return null;
    }

    /**
     * Transforms the string and returns the rendered value.
     * The value is modified using the arguments and an optional filterContext.
     * The filterContext holds "softened" version of the arguments that it can use
     * on multiple invocations for the sake of efficiency.  For instance in an "add"
     * filter that would increment the value of an input string by some amount, 
     * it makes sense to cache the value as an integer instead of converting the
     * stringified argument on each render call!
     * @throws IOException 
     */
    public Object apply(Object input, TemplateContext context, ArrayList<Variable>arguments)
    {
    	return null;
    }
}

File: objc/Classes/filters/Filter.m
//
//  Filter.m
//  DjangoTemplates
//
//  Created by Sri Panyam on 2/02/10.
//  Copyright 2010 __MyCompanyName__. All rights reserved.
//

#import "Filter.h"


@implementation Filter

@synthesize arguments;

/**
 * INitialises the filter with a name
 */
- (id) init
{
	if (self = [super init])
	{
		arguments = nil;
	}
	return self;
}

/**
 * Returns the name of the filter.
 * To be overridden by child classes.
 */
- (NSString *)filterName
{
    return @"";
}

/**
 * Applies the filter on a string and returns the transformed value.
 */
- (NSString *)render:(NSString *)value
{
	// by default we do nothing and return the original string
	return value;
}

/**
 * Cleans up the filter
 */
- (void) dealloc
{
	[arguments removeAllObjects];
	[super dealloc];
}

/**
 * Creates a filter instance specific context.
 * @return A filter specific context object.
 */
- (id)makeInstanceContext:(NSArray *)arguments
{
    return nil;
}

/**
 * Creates an invocation specific context for the filter.
 * @return An invocation specific context object for this filter.
 */
- (id)makeInvocationContext:(NSArray *) arguments
{
    return nil;
}

/**
 * Transforms the string and returns the rendered value.
 * The value is modified using the arguments and an optional filterContext.
 * The filterContext holds "softened" version of the arguments that it can use
 * on multiple invocations for the sake of efficiency.  For instance in an "add"
 * filter that would increment the value of an input string by some amount, 
 * it makes sense to cache the value as an integer instead of converting the
 * stringified argument on each render call!
 * @throws IOException 
 */
- (BOOL) render:(NSOutputStream *)output
           input:(NSString *)input
           arguments:(NSArray *)arguments
           filterContext:(NSObject *)filterContext
           error:(NSError  **)error
{
    return YES;
}

@end
